util/linuxfw/nftables_runner.go

```diff
@@ -0,0 +1,1058 @@
+			DestRegister: 1,
+			Base:         expr.PayloadBaseNetworkHeader,
+			Offset:       12,
+			Len:          4,
+		}, nil
+	case nftables.TableFamilyIPv6:
+		return &expr.Payload{
+			DestRegister: 1,
+			Base:         expr.PayloadBaseNetworkHeader,
+			Offset:       8,
+			Len:          16,
+		}, nil
+	default:
+		return nil, fmt.Errorf("Table family passed in is neither IPv4 nor IPv6")
+	}
+}
+
+// HasIPV6 returns true if the system supports IPv6.
+func (n *nftablesRunner) HasIPV6() bool {
+	return n.v6Available
+}
+
+// HasIPV6NAT returns true if the system supports IPv6 NAT.
+func (n *nftablesRunner) HasIPV6NAT() bool {
+	return n.v6NATAvailable
+}
+
+// findRule parse through the rules and find the rule with matching expression.
+// Matching expressions may not be in order.
+func findRule(conn *nftables.Conn, rule *nftables.Rule) (*nftables.Rule, error) {
+	rules, err := conn.GetRules(rule.Table, rule.Chain)
+	if err != nil {
+		return nil, fmt.Errorf("get nftables rules: %w", err)
+	}
+	if len(rules) == 0 {
+		return nil, nil
+	}
+
+	for _, r := range rules {
+		if len(r.Exprs) != len(rule.Exprs) {
+			continue
+		}
+		match := true
+		for i, e := range r.Exprs {
+			if !reflect.DeepEqual(e, rule.Exprs[i]) {
+				match = false
+				break
+			}
+		}
+		if match {
+			return r, nil
+		}
+	}
+
+	return nil, fmt.Errorf("The rule is not found in chain %q of table %q", rule.Chain.Name, rule.Table.Name)
+}
+
+// insertLoopbackRule inserts the TS loop back rule into the given chain as the first rule.
+func insertLoopbackRule(conn *nftables.Conn, proto nftables.TableFamily, table *nftables.Table, chain *nftables.Chain, addr netip.Addr) error {
+	matchingAddr := addr.AsSlice()
+
+	saddrExpr, err := newLoadSaddrExpr(proto)
+	if err != nil {
+		return fmt.Errorf("get expr: %w", err)
+	}
+	loopBackRule := &nftables.Rule{
+		Table: table,
+		Chain: chain,
+		Exprs: []expr.Any{
+			&expr.Meta{
+				Key:      expr.MetaKeyIIFNAME,
+				Register: 1,
+			},
+			&expr.Cmp{
+				Op:       expr.CmpOpEq,
+				Register: 1,
+				Data:     []byte("lo"),
+			},
+			saddrExpr,
+			&expr.Cmp{
+				Op:       expr.CmpOpEq,
+				Register: 1,
+				Data:     matchingAddr,
+			},
+			&expr.Counter{},
+			&expr.Verdict{
+				Kind: expr.VerdictAccept,
+			},
+		},
+	}
+
+	//This inserts rule to the top of the chain
+	_ = conn.InsertRule(loopBackRule)
+
+	if err = conn.Flush(); err != nil {
+		return fmt.Errorf("insert rule: %w", err)
+	}
+	return nil
+}
+
+// getNFTByAddr returns the nftables with correct IP family
+// that we will be using for the given address.
+func (n *nftablesRunner) getNFTByAddr(addr netip.Addr) *nftable {
+	if addr.Is6() {
+		return n.nft6
+	}
+	return n.nft4
+}
+
+// AddLoopbackRule adds an nftables rule to permit loopback traffic to
+// a local Tailscale IP.
+func (n *nftablesRunner) AddLoopbackRule(addr netip.Addr) error {
+	nf := n.getNFTByAddr(addr)
+
+	inputChain, err := getChainFromTable(n.conn, nf.Filter, tsChain(TSInputChain))
```

Comment:
I don't think we should do this for 2 reasons.

1. It will provide extra complexity, we shouldn't have this info just stored in the `nftable` struct because they shouldn't be stored at the same level as the tables since they are contained by tables. So we will need another layer of container representing a nftables.Table.
2. We don't have that many calls to this method and doing a call to get chains is not slow since we already have table name and family, and the table is our own table so there are maximum 2 chains in the table.

We can alternatively save chains instead of tables in `nftable`, since chain would contain table, and usually these two pieces of info are used together, but that would provide additional complexity and most of the code would have to be reworked.